tests/ui/underscore-imports/issue-110164.rs

```diff
@@ -0,0 +1,19 @@
+use self::*;
+//~^ ERROR unresolved import `self::*`
+use crate::*;
+//~^ ERROR unresolved import `crate::*`
+use _::a;
+//~^ ERROR expected identifier, found reserved identifier `_`
+//~| ERROR unresolved import `_`
+use _::*;
+//~^ ERROR expected identifier, found reserved identifier `_`
+//~| ERROR unresolved import `_`
+
+fn main() {
+    use _::a;
+    //~^ ERROR expected identifier, found reserved identifier `_`
+    //~| ERROR unresolved import `_`
+    use _::*;
```

Comment:
Why does `use _::*` trigger the "glob-import a module into itself" in the first place?
I'd expect `_` to resolve into `Res::Err` here.

Is it possible to reproduce the ICE without `_::*`, just with regular well-formed imports?

Reply:
> Is it possible to reproduce the ICE without _::*, just with regular well-formed imports?

Will not occur, other cases will hit [map_err](https://github.com/rust-lang/rust/blob/master/compiler/rustc_resolve/src/ident.rs#L870)

> I'd expect _ to resolve into Res::Err here.

When executing [new_key](https://github.com/rust-lang/rust/blob/master/compiler/rustc_resolve/src/ident.rs#L870), [special processing](https://github.com/rust-lang/rust/blob/master/compiler/rustc_resolve/src/lib.rs#L1594) is performed for `_` to ensure that `_` each time the `key` is is not the same, it is intended to solve the case of [`xxx as _`](https://github.com/rust-lang/rust/blob/master/tests/ui/underscore-imports/basic.rs#L33)

Reply:
Haha, fascinating.
The `underscore_disambiguator += 1` should only apply when we define names, not when we are retrieving already added definitions.

Reply:
I suggest adding `BindingKey::new` method that will be used everywhere by default, and rename `new_key` to `new_disambiguated_key` and use it only when adding `use foo as _` imports (in `fn add_import` I guess, but I'm not sure).

Note that if `new_key` is called for a single import twice in different contexts, then the resulting disambiguators are not consistent with each other, and it's something that is suspicious at least.

Reply:
This change is independent from glob imports and we should do it first.

The glob change also looks reasonable, maybe it fixes some other actual issues.

Reply:
Then I will open a new PR to rewrite `new_key`, what do you think?

Reply:
Yes, new PR.